dwave/plugins/torch/samplers/block_spin_sampler.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import TYPE_CHECKING, Callable, Hashable, Literal, Optional
+
+import torch
+from torch import nn
+
+if TYPE_CHECKING:
+    from dwave.plugins.torch.models.boltzmann_machine import GraphRestrictedBoltzmannMachine as GRBM
+
+from dwave.plugins.torch.nn.functional import bit2spin_soft
+from dwave.plugins.torch.tensor import randspin
+
+__all__ = ["BlockSpinSampler"]
+
+
+class BlockSpinSampler:
+    """A block-spin update sampler for graph-restricted Boltzmann machines.
+
+    Due to the sparse definition of GRBMs, some tedious indexing tricks are required to
+    efficiently sample in blocks of spins. Ideally, an adjacency list can be used, however,
+    adjacencies are ragged, making vectorization inapplicable.
+
+    Block-Gibbs and Block-Metropolis obey detailed balance and are ergodic methods at finite
```

Comment:
Block-Metropolis is not ergodic at zero temperature. Change to 'finite non-zero'